src/lib/blkalloc/append_blk_allocator.cpp

```diff
@@ -0,0 +1,180 @@
+        return BlkAllocStatus::FAILED;
+    }
+
+    // Push 1 blk to the vector which has all the requested nblks;
+    out_bids.emplace_back(m_last_append_offset, nblks, m_chunk_id);
+
+    m_last_append_offset += nblks;
+    m_freeable_nblks -= nblks;
+
+    set_dirty_offset();
+
+    COUNTER_INCREMENT(m_metrics, num_alloc, 1);
+
+    return BlkAllocStatus::SUCCESS;
+}
+
+//
+// Alternative design option:
+// Option-1. Everytime offset is updated in a single append_blk_allocator instance, update the dirty buffer into
+// VDevCPContext, which contains a list of (ThreadVector) dirty buffers from all the append_blk_allocator instances
+// (protected by RCU);
+//
+// And cp_flush flush every dirty buffer in that list to disk;
+//
+// Option-2. vdev cp_flush will call each individual chunk's blkallocator and each append_blk_allocator instance flush
+// its dirty buffer on demond;
+//
+// We are using option-2.
+//
+
+//
+// cp_flush is happening in different thread that is running concurrently as AppendBlkAllocator::alloc;
+//
+// 1. if clear_dirty_offset happens before offset is updated by alloc, but before set_dirty_offset, we are flushing
+// offset twice, which is fine.
+// 2. If clear_dirty_offset happens after set_dirty_offset, we are flushing offset ahead of log
+// store, which is also fine during recovery, it is consistent after log store flush completes;
+// 3. Any alloc happened after metablk is persisted will come to next cp;
+//
+void AppendBlkAllocator::cp_flush() {
+    if (m_is_dirty) {
+        // clear must happen before write to metablk becuase if metablk is written first, if
+        // alloc(in-parallel) happened after written but before clear, then the dirty buffer will be cleared.
+        clear_dirty_offset();
+
+        // write to metablk;
+        m_sb->last_append_offset = m_last_append_offset;
+        m_sb->freeable_nblks = m_freeable_nblks;
+        m_sb.write();
+    }
+}
+
+void AppendBlkAllocator::set_dirty_offset() { m_is_dirty = true; }
+
+void AppendBlkAllocator::clear_dirty_offset() { m_is_dirty = false; }
+
+//
+// free operation does:
+// 1. book keeping "total freeable" space
+// 2. if the blk being freed happens to be last block, move last_append_offset backwards accordingly;
+//
+void AppendBlkAllocator::free(const BlkId& bid) {
+    const auto n = bid.get_nblks();
+    m_freeable_nblks += n;
+    if (bid.get_blk_num() + n == m_last_append_offset) {
+        // we are freeing the the last blk id, let's rewind.
+        m_last_append_offset -= n;
+    }
+    set_dirty_offset();
+}
+
+void AppendBlkAllocator::free(const std::vector< BlkId >& blk_ids) {
+    for (const auto b : blk_ids) {
+        free(b);
+    }
+}
+
+blk_cap_t AppendBlkAllocator::available_blks() const { return get_total_blks() - get_used_blks(); }
+
+blk_cap_t AppendBlkAllocator::get_used_blks() const { return m_last_append_offset; }
+
+bool AppendBlkAllocator::is_blk_alloced(const BlkId& in_bid, bool use_lock) const {
```

Comment:
either [[maybe_unused]] for use_lock or better no declaration for use_lock like
```bool AppendBlkAllocator::is_blk_alloced(const BlkId& in_bid, bool) const {```

Reply:
done